### PR TITLE
♻️ Refactor Auth0 Blueprint & GitHub Service

### PR DESCRIPTION
**Pull request by jasonBirchall**
## 👀 Purpose

- This pull connects to https://github.com/orgs/ministryofjustice/projects/52/views/4?pane=issue&itemId=50422029

## ♻️ What's changed

- ✨  Created a new route in the "join" section that directs users to a placeholder template as a way to thank them. This is a preliminary step, with further modifications planned in upcoming PRs.

- ♻️ Overhauled the authentication route to follow Auth0's recommended blueprint approach. This refactoring aligns the application with best practices as outlined in Auth0's [documentation](https://auth0.com/docs/quickstart/webapp/python/interactive).

- 🔥  Removed unnecessary parts of the GitHub service, focusing solely on adding users to their selected organisation. This change makes the GitHub interaction more concise and relevant to the current private beta requirements.

- 🔥  Streamlined the Auth0 validation process to focus primarily on verifying if the user's domain is allowed and ensuring the user has authenticated with the same email they started with.

- 🔧  Introduced a new configuration file `join_github_app/main/config/github.py` to replace the `SEND_EMAIL_INVITES` constant. This change allows the configuration value to be set via an environment variable, enhancing flexibility and deployment ease.

- ✨ Added an AuthTokenError class to the error_handler module, providing more specific error handling during authentication.

- 🧪 Updated and refactored the views and authentication tests to reflect these changes, ensuring that the test suite remains reliable and in sync with the application's functionality.

## 📝 Notes

-

### CODE REVIEW

File: .github/workflows/cicd-deploy-to-prod.yml [hunk 1]
```diff
@@ -18,6 +18,7 @@ env:
   ADMIN_GITHUB_TOKEN: ${{ secrets.PROD_ADMIN_GITHUB_TOKEN }}
   FLASK_APP_SECRET: ${{ secrets.PROD_FLASK_APP_SECRET }}
   FLASK_CONFIGURATION: "production"
+  SEND_EMAIL_INVITES: "True"
```
Comment (connormaglynn) [inline .github/workflows/cicd-deploy-to-prod.yml:21]: I'd recommend leaving this as false until we've run through the security assessment with the MvP Journey - never know what else we might find 🔒 👀

Comment (connormaglynn) [reply .github/workflows/cicd-deploy-to-prod.yml:21]: We can temporarily turn it on in Dev to test, if needed 🧪

Comment (jasonBirchall) [reply .github/workflows/cicd-deploy-to-prod.yml]: Agreed, changed

File: tests/routes/test_auth.py [hunk 2]
```diff
@@ -2,113 +2,165 @@
...
+    @patch("join_github_app.main.routes.auth.get_token")
+    def test_callback_route_token_failure(self, mock_get_token, mock_render_error_page):
+        mock_get_token.side_effect = AuthTokenError("Token error")
+
+        response = self.client.get("/auth/callback")
+        self.assertEqual(response.status_code, 500)
+        mock_get_token.assert_called_once()
+        mock_render_error_page.assert_called_once()
+
+    @patch("join_github_app.main.routes.auth.ALLOWED_EMAIL_DOMAINS",
```
Comment (tamsinforbes) [inline tests/routes/test_auth.py:80]:
```suggestion
    @patch("join_github_app.main.routes.auth.ALLOWED_EMAIL_DOMAINS",
```
Presumably this doesn't need a `clear=True` arg as `@patch.dict` does?

File: tests/routes/test_auth.py [hunk 3]
```diff
@@ -2,113 +2,165 @@
...
+            session['user'] = {'userinfo': {'email': 'user@test.com'}}
+            session['email'] = 'user@test.com'
+            session['org_selection'] = ['some_org']
+
+            result = process_user_session()
+            self.assertFalse(result)
+
+    @patch("join_github_app.main.routes.auth.ALLOWED_EMAIL_DOMAINS",
+           new=["example.com", "test.com"])
+    def test_process_user_session_with_invalid_user(self):
```
Comment (tamsinforbes) [inline tests/routes/test_auth.py:111]:
```suggestion
    def test_process_user_session_with_mismatched_user_email(self):
```

File: tests/routes/test_auth.py [hunk 4]
```diff
@@ -2,113 +2,165 @@
...
+
+            result = process_user_session()
+            self.assertFalse(result)
+
+    @patch("join_github_app.main.routes.auth.ALLOWED_EMAIL_DOMAINS",
+           new=["example.com", "test.com"])
+    def test_process_user_session_with_missing_org_selection(self):
+        with self.app.test_request_context('/auth/callback'):
+            session['user'] = {'userinfo': {'email': 'user@test.com'}}
+            session['email'] = 'user@test2.com'
```
Comment (tamsinforbes) [inline tests/routes/test_auth.py:125]:
```suggestion
            session['email'] = 'user@test.com'
```
Copy/paste/edit fail

File: tests/routes/test_auth.py [hunk 5]
```diff
@@ -2,113 +2,165 @@
...
+            session['org_selection'] = ['some_org']
```
Comment (tamsinforbes) [inline tests/routes/test_auth.py:126]:
```suggestion
            session['org_selection'] = []
```
Missing org selection?

**Review by tamsinforbes — CHANGES_REQUESTED** [review]
💥Looks great, couple of Qs/typos

**Review by tamsinforbes — APPROVED** [review]
LGTM 🚀

**Pull request merged**